Feature/us11/envoie donnees mesure

JarodEphec opened the pull request:
# Présentation de la pull request

Cette US a pour but de créer les routes API et les schémas de DB qui permettent : 

- La création et la lecture des aquariums dans la DB pour un utilisateur enregistré.
- La création et la lecture des diagnostics dans la DB pour un aquarium.
- L'enregistrement et la lecture des mesures effectuées par une sonde dans la DB pour un aquarium.

Toutes ces routes requièrent l'authentification de l'utilisateur sur base du modèle créé avec l'US 07.

Elle est surement la première itération de l'US 11, cette US va peut-être évoluer en fonction des besoins et de la façon dont les données vont être envoyées à la DB. Description de cette US peut être trouvée [ici](https://github.com/users/Martin-Hayot/projects/2?pane=issue&itemId=40037408).

# Utilisation de l'API

Voici un exemple de l'utilisation de l'API.

## Aquarium

### POST

URL : 
```
/api/aquarium
```
Body : 
```
{
    name: "Mon aquarium de 2m²",
}
```

### GET

URL : 
```
/api/aquarium
```
headers : 
```
headers: {
  cookies : [
    accessToken: accessToken,
    refreshToken: refreshToken
  ]
}
```

Réponse : 
```
{
    "_id":"6543a8f911cd8288c7143d02",
    "userId":"6533f52718ac173412f5cb86",
    "name":"Mon aquarium de 2m²",
}
```
## Diagnostic

### POST

URL : 
```
/api/aquarium/diagnostic
```
Body : 
```
{
    aquariumId: "6543a8f911cd8288c7143d02",
    date: "2023-11-02T15:36:48.384Z",
}
```

### GET

URL : 
```
/api/aquarium/:aquariumId/diagnostic
```
Parameter : 
```
/api/aquarium/654f71941fdc527c68052cc0/diagnostic
```

Réponse : 
```
{
    "_id": "6543b6d060921eeaaea1f833",
    "aquariumId": "6543a8f911cd8288c7143d02",
    "date": "2023-11-02T15:36:48.384+00:00",
}
```

## Data

### POST

URL : 
```
/api/aquarium/diagnostic/data
```
Body : 
```
{
    diagnosticId: 'hfoia97bsocvo1629hfdqsvf',
    measure: 1200,
    frequency: 200,
}
```

### GET

URL : 
```
/api/aquarium/:aquariumId/diagnostic/:diagnosticId/data
```
Parameter : 
```
/api/aquarium/654f71941fdc527c68052cc0/diagnostic/654f7ad36941701120cea607/data
```

Réponse : 
```
{
    "_id": "6543c215181a47e33cadb4fb",
    "diagnosticId": "6543b6d060921eeaaea1f833",
    "measure": 51,
    "frequency": 200,
}
```

Kouvahe81 commented:
Je vois que tu n'as pas ajouter le ou les clés étrangères comme sur le schma de la DB
Exemple:
const templateReservation = new mongoose.Schema({
dateResevation:{
        type: Date,
        required: true
    },
    clients:[{type:mongoose.Schema.Types.ObjectId, ref:'Client'}],
    festivalDays:[{type: mongoose.Schema.Types.ObjectId, ref:'FestivalDay'}],
})
Sur cette table clients et festivalDays sont des clés étrangères. Tu peux t'en inspirer.

JarodEphec commented:
> Je vois que tu n'as pas ajouter le ou les clés étrangères comme sur le schma de la DB Exemple: const templateReservation = new mongoose.Schema({ dateResevation:{ type: Date, required: true }, clients:[{type:mongoose.Schema.Types.ObjectId, ref:'Client'}], festivalDays:[{type: mongoose.Schema.Types.ObjectId, ref:'FestivalDay'}], }) Sur cette table clients et festivalDays sont des clés étrangères. Tu peux t'en inspirer.

Ah ok, on peut faire des clés étrangères en mongoDB, je ne savais pas, merci :)

Martin-Hayot commented on backend/routes/aquarium.js:
```diff
@@ -0,0 +1,23 @@
+const { requireUser } = require("../middleware/auth");
+const Aquarium = require("../models/sonde/aquarium");
+
+router.post("/", requireUser, (req, res, next) => {
+    const aquarium = new Aquarium({
+        ...req.body,
+    });
+    aquarium
+        .save()
+        .then(() => res.status(201).json({ message: "Data enregistré !" }))
+        .catch((error) => res.status(400).json({ error }));
+});
```
Je ne ferais pas comme ça ici. 
Je m'explique, l'authentification que j'ai fait créer un objet user dans l'objet req donc le userId ne devraient pas se trouver dans le body. D'ailleurs pour le model jsp hn mais ce serait pas mieux de faire avec l'email ? c'est plus simple à gérer en général.
Et faire une déstructuration est pas une bonne idée parce que tu ne vérifie pas ce qui il y a dans le body.
Encore un point à modifier c'est le next dans le callback, tu dois pas le mettre. le next doit seulement se trouver dans un middleware.

je ferais donc plutôt comme ça : 

```js
router.post("/", requireUser, async (req, res) => {
	const { userId, email } = req.user;
	const { data, name } = req.body; // ici tu viens récupérer les éléments utiles pour la req
	if (!data || !name) {
		return res.status(400).json({ message: "All entries are required" });
	}
	const aquarium = new Aquarium({
		userId,
		data,
		name,
	});
	try {
		await aquarium.save();
		res.status(201).json({ message: "Aquarium created successfully" });
	} catch (e) {
		console.log("Error : ", e);
		res.status(500).json({ message: e });
	}
});
```

Martin-Hayot commented on backend/routes/aquarium.js:
```diff
@@ -0,0 +1,23 @@
+        .catch((error) => res.status(400).json({ error }));
+});
+
+router.get("/", requireUser, (req, res, next) => {
+    const { email } = req.user;
+    Aquarium.find({ userId: email })
+        .then((aquarium) => res.status(200).json(aquarium))
+        .catch((error) => res.status(400).json({ error }));
+});
```
fait attention ici que le userId n'est pas l'email !. Et de nouveau les erreurs sont mal gérer. Le code est trop générique et ne correspond pas aux différentes erreurs qu'il pourrait y avoir. 
ex: 
code 500: erreur serveur
code 404: not found
etc...

JarodEphec replied:
Ici, je m'étais dit que ça serait intéressant d'utiliser l'email de l'utilisateur comme l'id de ce dernier.

Martin-Hayot replied:
Ok mais le problème c'est que userId correspond à l'id d'un utilisateur dans le model User. userId étant à la base reçu dans le req.user grâce à l'authentification..
```js
	const {
		email, // email de l'utilisateur
		userId, // id de l'utilisateur
	} = req.user;
	const searchedUser = await User.findById({ _id: userId }); // _id est un champ de la base de données créé automatiquement par MongoDB
```
donc le fait que userId soit l'email porte à confusion même si je comprends bien le fait que c'est un identifiant unique.
mais ducoup je rejoint aussi Serge la dessus, le userId devraient être une clé étrangère pointant sur le bonne id de la table User

Martin-Hayot reviewed (changes requested):
Pour le reste du code, les commentaires que j'ai à faire reste les mêmes. 

Donc je résume : 
- ne pas directement déstructurer sans réel intérêts et faire de check d'abord pour éviter des erreurs imprévues
- utiliser l'objet req.user pour récupérer des informations sur l'utilisateur
- mieux gérer les erreurs et leurs code http et ne pas oublier de stocker le message dans un objet json
- pas mettre de paramètre next dans le callback

Martin-advinci reviewed (changes requested):
Juste encore une remarque sinon le reste c'est parfait ! 
Pour que l'api suivent bien l'architecture rest. Il faudrait que les routes ressemblent plutôt à ça: 

```js
route.get("/:id/diagnostic") // pour récupérer tout les diagnostic d'un aquarium
route.get("/:id/diagnostic/:id") // pour récupérer un diagnostic d'un aquarium
```

Martin-Hayot reviewed (approved):
Parfait !

JarodEphec committed:
US 11 feature: Ajout des routes restfull api, pour les aquariums.

Cette US a pour but de créer les routes API et les schémas de DB qui permettent :

- La création et la lecture des aquariums dans la DB pour un utilisateur enregistré.
- La création et la lecture des diagnostics dans la DB pour un aquarium.
- L'enregistrement et la lecture des mesures effectuées par une sonde dans la DB pour un aquarium.

Toutes ces routes requièrent l'authentification de l'utilisateur sur base du modèle créé avec l'US 07.

Elle est surement la première itération de l'US 11, cette US va peut-être évoluer en fonction des besoins et de la façon dont les données vont être envoyées à la DB. Description de cette US peut être trouvée [ici](https://github.com/users/Martin-Hayot/projects/2?pane=issue&itemId=40037408).

# Utilisation de l'API

Voici un exemple de l'utilisation de l'API.

## Aquarium

### POST

URL :
```
/api/aquarium
```
Body :
```
{
    name: "Mon aquarium de 2m²",
}
```

### GET

URL :
```
/api/aquarium
```
headers :
```
headers: {
  cookies : [
    accessToken: accessToken,
    refreshToken: refreshToken
  ]
}
```

Réponse :
```
{
    "_id":"6543a8f911cd8288c7143d02",
    "userId":"6533f52718ac173412f5cb86",
    "name":"Mon aquarium de 2m²",
}
```
## Diagnostic

### POST

URL :
```
/api/aquarium/diagnostic
```
Body :
```
{
    aquariumId: "6543a8f911cd8288c7143d02",
    date: "2023-11-02T15:36:48.384Z",
}
```

### GET

URL :
```
/api/aquarium/:aquariumId/diagnostic
```
Parameter :
```
/api/aquarium/654f71941fdc527c68052cc0/diagnostic
```

Réponse :
```
{
    "_id": "6543b6d060921eeaaea1f833",
    "aquariumId": "6543a8f911cd8288c7143d02",
    "date": "2023-11-02T15:36:48.384+00:00",
}
```

## Data

### POST

URL :
```
/api/aquarium/diagnostic/data
```
Body :
```
{
    diagnosticId: 'hfoia97bsocvo1629hfdqsvf',
    measure: 1200,
    frequency: 200,
}
```

### GET

URL :
```
/api/aquarium/:aquariumId/diagnostic/:diagnosticId/data
```
Parameter :
```
/api/aquarium/654f71941fdc527c68052cc0/diagnostic/654f7ad36941701120cea607/data
```

Réponse :
```
{
    "_id": "6543c215181a47e33cadb4fb",
    "diagnosticId": "6543b6d060921eeaaea1f833",
    "measure": 51,
    "frequency": 200,
}
```
